packages/cli/src/typescript/project-cache.ts

```diff
@@ -0,0 +1,152 @@
+  const projects: Record<string, Partial<Record<AllPlatforms, Project>>> = {};
+
+  function findProjectRoot(sourceFile: string): string {
+    // Search known root directories to see if the source file is in one of them.
+    for (const root of Object.keys(projects)) {
+      if (sourceFile.startsWith(root)) {
+        return root;
+      }
+    }
+
+    // Search the file system for the root of source file's package.
+    const root = findPackageDir(path.dirname(sourceFile));
+    if (!root) {
+      throw new Error(
+        `Cannot find project root for source file '${sourceFile}'`
+      );
+    }
+
+    return root;
+  }
+
+  function readTSConfig(root: string): ts.ParsedCommandLine {
+    const configFileName = path.join(root, "tsconfig.json");
+
+    const cmdLine = readConfigFile(configFileName);
+    if (!cmdLine) {
+      throw new Error(`Failed to load '${configFileName}'`);
+    } else if (cmdLine.errors.length > 0) {
+      const writer = createDiagnosticWriter();
+      cmdLine.errors.forEach((e) => writer.print(e));
+      throw new Error(`Failed to load '${configFileName}'`);
+    }
+
+    return cmdLine;
+  }
+
+  function createProject(root: string, platform: AllPlatforms): Project {
+    // Load the TypeScript configuration file for this project.
+    const cmdLine = readTSConfig(root);
+
+    // Create a TypeScript project using the configuration file. Enhance the
+    // underlying TS language service with our react-native module resolver.
+    const enhanceLanguageServiceHost = (host: ts.LanguageServiceHost): void => {
+      const platformExtensionNames = platformExtensions(platform);
+      const disableReactNativePackageSubstitution = true;
+      const traceReactNativeModuleResolutionErrors = false;
+      const traceResolutionLog = undefined;
+      changeHostToUseReactNativeResolver({
+        host,
+        options: cmdLine.options,
+        platform,
+        platformExtensionNames,
+        disableReactNativePackageSubstitution,
+        traceReactNativeModuleResolutionErrors,
+        traceResolutionLog,
+      });
+    };
+    const tsproject = new Project(
+      documentRegistry,
+      diagnosticWriter,
+      cmdLine,
+      enhanceLanguageServiceHost
+    );
+
+    // Start with an empty project, ignoring the file graph from tsconfig.json.
+    tsproject.removeAllFiles();
+
+    return tsproject;
+  }
+
+  function getProject(sourceFile: string, platform: AllPlatforms): Project {
+    const root = findProjectRoot(sourceFile);
+    if (!projects[root]) {
+      projects[root] = {};
+    }
+    if (!projects[root][platform]) {
+      projects[root][platform] = createProject(root, platform);
+    }
+    return projects[root][platform]!;
+  }
+
+  function clearPlatform(platform: AllPlatforms): void {
+    Object.values(projects).forEach((projectsByPlatform) => {
+      if (projectsByPlatform[platform]) {
+        projectsByPlatform[platform]!.dispose();
+        delete projectsByPlatform[platform];
+      }
```

Comment:
Nit: Can we store a temp variable to avoid the nonnull assertion?
```suggestion
      const project = projectsByPlatform[platform]
      if (project) {
        project.dispose();
        delete projectsByPlatform[platform];
      }
```

Reply:
Fixed